src/plugin.rs

```diff
@@ -678,34 +651,15 @@ pub trait Plugin {
     /// Handle incoming events sent from the host.
     ///
     /// This is always called before the start of `process` or `process_f64`.
+    ///
+    /// This method is only called while the plugin is in the *resumed* state.
     fn process_events(&mut self, events: &api::Events) {}
 
-    /// Return handle to plugin editor if supported.
-    fn get_editor(&mut self) -> Option<&mut Editor> {
-        None
-    }
-
-
-    /// If `preset_chunks` is set to true in plugin info, this should return the raw chunk data for
-    /// the current preset.
-    fn get_preset_data(&mut self) -> Vec<u8> {
-        Vec::new()
+    /// Get a reference to the shared parameter object.
+    fn get_parameter_object(&mut self) -> Arc<PluginParameters> {
```

Comment:
Two remarks here:
1. We are using trait objects here. I see two options here:
a. Keep using trait objects, but make it explicit by using the `dyn` keyword (`Arc<dyn PluginParameters>`)
b. Switch to normal generics by defining an extra associated type `type PluginParameterType: PluginParameters;` (temporary placeholder name) to the trait and changing the return type to `Arc<Self::PluginParameterType`. Unfortunately, if we do this, we cannot specify a default associated type here and hence we cannot give a default implementation (see rust-lang/rust#29661, also not possible with a generic trait `trait Plugin<PluginParameterType: PluginParameters>` instead of associated types). This means that all plugins, even the ones that don't use parameters, need to write some dummy code. 
So we have to make a choice here: either "don't pay a run-time overhead for dynamic dispatch if you don't use it" or "don't write extra code for parameters if your plugin does not have parameters". I think the run-time overhead of dynamic dispatch (trait objects) is rather small in this context, so I can understand the choice for dynamic dispatch here.
2. As far as I know, there are two mechanisms one can use for concurrency in this context: atomics (as you use in the example), or some lock-free message passing such as the crossbeam-channel crate. As far as I understood, crossbeam's `Sender`s do not need to be put inside an `Arc` in order to work as intended (in contrast to atomics). Hence the use of `Arc` is opinionated. I think we can change the signature to something like `fn get_parameter_object(&mut self) -> Box<dyn PluginParameters>` or `fn get_parameter_object(&mut self) -> Self::PluginParameterType` and then later on `impl PluginParameters for Arc<DummyPluginParameters> {}`. This has the disadvantage that most people who are new to Rust are not used to writing `impl Trait for Arc<MyType>`, which can be confusing, and also the orphan rules may give some trouble. Given these considerations, I think we can also acknowledge that there is no ideal solution and keep it as is, but it depends on how high-level or low-level we see this crate.

Reply:
Thank you for the review and the deep-diving comments. I will try to address each of your points in the following:

I agree that we should use the `dyn` syntax for trait objects. The syntax was not yet stable when I opened the PR. Trait objects are used in many placed throughout the crate (also in places not related to this change), so I think it makes most sense to have a separate commit that changes all of them.

Adding an associated type will make the `Plugin` trait not object safe. The `Plugin` trait is used as a trait object in the glue code all over the crate (constructed [here](https://github.com/rust-dsp/rust-vst/blob/master/src/api.rs#L154)). Changing this will require many parts to become generic, and possibly a redesign of the way the plugin is initialized in the `plugin_main!` macro. It also would preclude the pure-Rust use case (using a Rust plugin directly in a Rust host).

So compared to the current situation, it is not an extra dynamic dispatch. It's just a dynamic dispatch to a different object. The dispatch is well predicted (always to the same target for a particular plugin) and happens once per host/plugin interaction. Indeed the overhead will be negligible.

Any synchronization mechanism that permits a shared end point can be used. For realtime-critical plugins, it is typically an additional requirement that the mechanism is lock-free and allocation-free, which further limits the choices available. Atomics and channels are examples, but I think it could make sense in a support crate to create more specialized mechanisms for this use case (on top of atomics, naturally).

[`crossbeam::channel::Sender`](https://docs.rs/crossbeam/0.6.0/crossbeam/channel/struct.Sender.html) is `Sync`, so it can be put inside the `Arc`. So is [`std::sync::mpsc::SyncSender`](https://doc.rust-lang.org/std/sync/mpsc/struct.SyncSender.html), but [`std::sync::mpsc::Sender`](https://doc.rust-lang.org/std/sync/mpsc/struct.Sender.html) is not.

It is not `Arc` that puts the `Sync` requirement onto the communication mechanism. It is simply the condition that the data must be shared across threads. This is implied by the fact that some of the methods can be called by multiple threads concurrently. Thus the requirement is unavoidable: Even if a different sharing mechanism than `Arc` was used, the `Sync` requirement would still be there.

One restriction that the use of `Arc` does impose on the plugin is that it requires the parameter object to be allocated on the heap. Since this happens once at plugin initialization, where a bunch of other objects are allocated on the heap anyway, I would not consider this a serious limitation in practice.

It would be nice to abstract over the kind of sharing mechanism used. However, Rust currently does not provide feasible ways of abstracting over type constructors (especially not in traits that need to be object safe). Also, `impl PluginParameters for Arc<MyType>` would not be permitted by the current orphan rules.

Does this address your concerns?

Reply:
Yes, this addresses all of my concerns.

Thanks for the detailed response! I learned some new things :-) and now I understand the design decisions better.